bitcoinspv/block_events.go

```diff
@@ -53,93 +53,70 @@ func (r *Relayer) handleBlockEvent(blockEvent *btctypes.BlockEvent) error {
 // onConnectedBlock handles connected blocks from the BTC client.
 // It is invoked when a new connected block is received from the Bitcoin node.
 func (r *Relayer) onConnectedBlock(blockEvent *btctypes.BlockEvent) error {
-	if err := r.validateBlockHeight(blockEvent); err != nil {
-		return err
-	}
-
-	if err := r.validateBlockConsistency(blockEvent); err != nil {
-		return err
-	}
-
-	ib, err := r.getAndValidateBlock(blockEvent)
-	if err != nil {
+	if err := r.ensureBlockConsistencyWithCache(blockEvent); err != nil {
 		return err
 	}
 
+	// new index block depend on Walrus config
+	ib := new(types.IndexedBlock)
 	// Store full block in Walrus
 	if r.Config.StoreBlocksInWalrus && r.walrusHandler != nil {
+		h := blockEvent.BlockHeader.BlockHash()
+		ib, err := r.btcClient.GetBTCBlockByHash(&h)
+		if err != nil {
+			return err
+		}
 		r.UploadToWalrus(ib.MsgBlock, ib.BlockHeight, ib.BlockHash().String())
+	} else {
+		ib.BlockHeight = blockEvent.Height
+		ib.MsgBlock.Header = *blockEvent.BlockHeader
 	}
 
-	r.btcCache.Add(ib)
+	err := r.btcCache.Add(ib)
+	if err != nil {
+		return err
+	}
 
 	return r.processBlock(ib)
 }
 
-func (r *Relayer) validateBlockHeight(blockEvent *btctypes.BlockEvent) error {
-	latestCachedBlock := r.btcCache.First()
-	if latestCachedBlock == nil {
-		err := fmt.Errorf("cache is empty, restart bootstrap process")
-		return err
+func (r *Relayer) ensureBlockConsistencyWithCache(b *btctypes.BlockEvent) error {
```

Comment:
lets add a comment what this function does step by step and what will happen in different scenario. This is the most important fucntion for handling the cache and reorgs. Please add a task to unit test it (you can assign me 🫠 )